File: midterm/summaryReport.js
const fs = require('fs');
const readline = require('readline');

// JavaScript (Node.js) implementation of the same task:
// - prompt for dataset path using readline
// - validate with fs.existsSync
// - compute simple analytics and write summary_report.csv using fs.writeFile

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function askFilePath() {
    rl.question('Enter dataset file path: ', function(path) {
        if (fs.existsSync(path)) {
            console.log('File found. Processing...');
            generateSummary(path);
        } else {
            console.log('Invalid file path. Try again.');
            askFilePath();
        }
    });
}

function generateSummary(path) {
    const data = fs.readFileSync(path, 'utf8');
    const lines = data.trim().split(/\r?\n/);

    let lineCount = Math.max(0, lines.length - 1); // skip header if present
    let sum = 0;
    let numValues = 0;

    for (let i = 1; i < lines.length; i++) {
        const parts = lines[i].split(',');
        for (const p of parts) {
            const v = parseFloat(p);
            if (!isNaN(v)) {
                sum += v;
                numValues++;
            }
        }
    }

    const outputLines = [
        'Metric,Value',
        `Lines,${lineCount}`,
        `NumericValues,${numValues}`,
        `Sum,${sum}`,
        `Average,${numValues > 0 ? sum / numValues : 0}`
    ];

    fs.writeFile('summary_report.csv', outputLines.join('\n'), (err) => {
        if (err) {
            console.error('Error writing summary:', err);
        } else {
            console.log('Summary written to summary_report.csv');
        }
        rl.close();
    });
}

askFilePath();

File: midterm/SummaryReport.java
import java.io.*;
import java.util.Scanner;

/**
 * Simple utility that reads a CSV dataset, computes a few sample analytics,
 * and writes a summary report to <code>summary_report.csv</code>.
 *
 * Requirements from MANAYSAY task:
 *  - prompt for file path using Scanner
 *  - validate using File class, loop until valid
 *  - perform analytics and export results using FileWriter
 */
public class SummaryReport {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        File dataFile;

        // prompt until a valid file is provided
        while (true) {
            System.out.print("Enter dataset file path: ");
            String path = input.nextLine();
            dataFile = new File(path);
            if (dataFile.exists() && dataFile.isFile()) {
                break;
            } else {
                System.out.println("Invalid file path. Please try again.");
            }
        }

        int lineCount = 0;
        double sum = 0;
        int numValues = 0;

        // read the CSV and do some simple analytics
        try (BufferedReader br = new BufferedReader(new FileReader(dataFile))) {
            // assume first line is header and skip it
            String header = br.readLine();
            String line;
            while ((line = br.readLine()) != null) {
                lineCount++;
                String[] parts = line.split(",");
                for (String part : parts) {
                    try {
                        double v = Double.parseDouble(part.trim());
                        sum += v;
                        numValues++;
                    } catch (NumberFormatException nfe) {
                        // ignore non‑numeric values
                    }
                }
            }
        } catch (IOException e) {
            System.err.println("Error reading data file: " + e.getMessage());
            System.exit(1);
        }

        // write the summary to CSV with headers
        try (FileWriter writer = new FileWriter("summary_report.csv")) {
            writer.append("Metric,Value\n");
            writer.append("Lines," + lineCount + "\n");
            if (numValues > 0) {
                writer.append("NumericValues," + numValues + "\n");
                writer.append("Sum," + sum + "\n");
                writer.append("Average," + (sum / numValues) + "\n");
            }
            System.out.println("Summary written to summary_report.csv");
        } catch (IOException e) {
            System.err.println("Error writing summary: " + e.getMessage());
        }
    }
}
